### PR TITLE
Start building the replacement task runner for Task Execution SDK

### PR DESCRIPTION
**Pull request by ashb**
The eventual goal of this "airflow.sdk.exeuction_time" package is to replace
LocalTaskJob and StandardTaskRunner, but at this stage it co-exists with it's
replacement.

As this PR is not a complete re-implementation of all the features that exist
currently (no handling of task level callbacks yet, no AirflowSkipException
etc.) the current tests are skeleton at best. Once we get closer to feature
parity (in future PRs) the tests will grow to match.

This supervisor and task runner operates slightly differently to the current
classes in these ways

**Logs from the subprocess are send over a different channel to stdout/stderr**

This makes the task supervisor a little bit more complex as it now has to
read stdout, stderr and a logs channel. The advantage of this approach is
that it makes the logs setup in the task process itself markedly simpler --
all it has to do is write logs output to the custom file handle as JSON and
it will show up "natively" as logs.

structlog has been chosen as the logging engine over stdlib's own logging as
the ability to have structured fields in the logs is nice, and stdlib is
configured to send logs to a stuctlog processor.

**Direct database access is replaced with an HTTP API client**

This is the crux of this feature and of AIP-72 in general -- tasks run via
this runner can no longer access DB models or DB session directly. This PR
doesn't yet implement the code/shims to make `Connection.get_connection_from_secrets`
use this client yet - that will be future work.

The reason tasks don't speak directly to the API server is primarily for two
reasons:

1. The supervisor process already needs to maintain an http session in order
   to report the task as started, to heart beat it, and to mark it as
   finished; and so because of that
2. Reduce the number of active HTTP connections for tasks to 1 per task
   (instead of 2 per task).

THe other reason we have this interface is that DAG parsing code will very
soon need to be updated to not have direct DB access either, and having this
"in process" interface ability already means that we can support commands like
`airflow dags reserialize` without having a running API server.

The API client itself is not auto-generated: I tried a number of different
client generates based on the OpenAPI spec and found them all lacking or buggy
in different ways, and the http client side itself is very simple, the only
interesting/difficult bit is the generation of the datamodels from the OpenAPI
spec which I found one that

msgspec was chosen over Pydantic as it is much lighter weight (and thus
quicker), especially on a client side when we have next to no validation
requirements of response data. I admit that I have not benchmarked it
specifically though.

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: task_sdk/pyproject.toml [hunk 1]
```diff
@@ -24,7 +24,11 @@ requires-python = ">=3.9, <3.13"
 dependencies = [
     "attrs>=24.2.0",
     "google-re2>=1.1.20240702",
+    "httpx>=0.27.0",
     "methodtools>=0.4.7",
+    "msgspec>=0.18.6",
```
Comment (jscheffl) [inline task_sdk/pyproject.toml:29]: If this is 0.18.6... can we assume it is stable and not having breaking changes by tomorrow?

File: task_sdk/src/airflow/sdk/api/client.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+
+    def auth_flow(self, request: httpx.Request):
+        if self.token:
+            request.headers["Authorization"] = "Bearer " + self.token
+        yield request
+
+
+def noop_handler(request: httpx.Request) -> httpx.Response:
+    log.debug("Dry-run request", method=request.method, path=request.url.path)
+    return httpx.Response(200, json={"text": "Hello, world!"})
```
Comment (jscheffl) [inline task_sdk/src/airflow/sdk/api/client.py:139]: What is this for?

File: task_sdk/src/airflow/sdk/api/client.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+def get_json_error(response: httpx.Response):
+    """Raise a ServerResponseError if we can extract error info from the error."""
+    err = ServerResponseError.from_response(response)
+    if err:
+        log.warning("Server error", detail=err.detail)
+        raise err
+
+
+def raise_on_4xx_5xx(response: httpx.Response):
+    return get_json_error(response) or response.raise_for_status()
```
Comment (jscheffl) [inline task_sdk/src/airflow/sdk/api/client.py:66]: Does it mean this API client does not implement a retry within?

File: task_sdk/src/airflow/sdk/api/datamodels/_generated.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+class TITargetStatePayload(Struct):
+    """
+    Schema for updating TaskInstance to a target state, excluding terminal and running states.
+    """
+
+    state: State
+
+
+class State1(Enum):
```
Comment (jscheffl) [inline task_sdk/src/airflow/sdk/api/datamodels/_generated.py:87]: What is the difference to `State`?

File: task_sdk/src/airflow/sdk/api/datamodels/_generated.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+class TIHeartbeatInfo(Struct):
+    """
+    Schema for TaskInstance heartbeat endpoint.
+    """
+
+    hostname: str
+    pid: int
+
+
+class State(Enum):
```
Comment (jscheffl) [inline task_sdk/src/airflow/sdk/api/datamodels/_generated.py:67]: What is the difference `TaskInstanceState` below?

File: task_sdk/src/airflow/sdk/api/datamodels/activities.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+import os
+
+import msgspec
+
+from airflow.sdk.api.datamodels.ti import TaskInstance
+
+
+class ExecuteTaskActivity(msgspec.Struct, tag="ExecuteTask", tag_field="kind"):
+    ti: TaskInstance
+    path: os.PathLike[str]
```
Comment (jscheffl) [inline task_sdk/src/airflow/sdk/api/datamodels/activities.py:29]: What is this `path` - the working dir of the task where it is called?

File: task_sdk/src/airflow/sdk/api/datamodels/ti.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+import uuid
+
+import msgspec
+
+
+class TaskInstance(msgspec.Struct, omit_defaults=True):
+    id: uuid.UUID
+
+    task_id: str
+    dag_id: str
```
Comment (jscheffl) [inline task_sdk/src/airflow/sdk/api/datamodels/ti.py:29]: Is some DAG version needed here as well?

File: task_sdk/src/airflow/sdk/execution_time/comms.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+r"""
+Communication protocol between the Supervisor and the task process
+==================================================================
+
+* All communication is done over stdout/stdin in the form of "JSON lines" (each
+  message is a single JSON document terminated by `\n` character)
+* Messages from the subprocess are all log messages and are sent directly to the log
+* No messages are sent to task process except in response to a request. (This is because the task process will
+  be running user's code, so we can't read from stdin until we enter our code, such as when requesting an XCom
+  value etc.)
```
Comment (jscheffl) [inline task_sdk/src/airflow/sdk/execution_time/comms.py:27]: Means that HTTP calls are synchronous and no async is possible? Else you would potentially mix responses and need to multiplex traffic?

Comment (jscheffl) [reply task_sdk/src/airflow/sdk/execution_time/comms.py:27]: Would it be better to call the Suporvisor being a HTTP bridge or proxy to the backend?

File: task_sdk/src/airflow/sdk/execution_time/comms.py [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+The reason this communication protocol exists, rather than the task process speaking directly to the Task
+Execution API server is because:
+
+1. To reduce the number of concurrent HTTP connections on the API server.
+
+   The supervisor already has to speak to that to heartbeat the running Task, so having the task speak to its
+   parent process and having all API traffic go through that means that the number of HTTP connections is
+   "halved". (Not every task will make API calls, so it's not always halved, but it is reduced.)
+
+2. This means that the user Task code doesn't ever directly see the task identity JWT token.
```
Comment (jscheffl) [inline task_sdk/src/airflow/sdk/execution_time/comms.py:38]: 3: Also al connection retry and error handling is being handled in the supervisor, so a task executed does not need to care about tćonnection flakiness?

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py
Comment (jscheffl) [inline task_sdk/src/airflow/sdk/execution_time/supervisor.py:1]: What is the difference between the "Supervisor" and the "Worker" like today CeleryWorker or EdgeWorker? Will the Supervisor replace these? Or will there be exactly 1 Supervisor as wrapper per task being executed?
(So if Celery executes with concurrency of 16, will to spawn 16 Supervisor or one that hosts 16 tasks?)

(And the question is mainly will each have it's own HTTP session and JWT or will thse calls be multiplexed?)

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py [hunk 10]
```diff
@@ -0,0 +1,547 @@
...
+                log.exception("Unable to decode message", line=line)
+                continue
+
+            # if isinstnace(msg, TaskState):
+            #     self._terminal_state = msg.state
+            # elif isinstance(msg, ReadXCom):
+            #     resp = XComResponse(key="secret", value=True)
+            #     encoder.encode_into(resp, buffer)
+            #     self.stdin.write(buffer + b"\n")
+            if isinstance(msg, GetConnection):
```
Comment (jscheffl) [inline task_sdk/src/airflow/sdk/execution_time/supervisor.py:471]: All that communication is very specifically implemented. Have you/we considered using a generic communication protocol like gRPC instead of writing our own?

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py
Comment (jscheffl) [reply task_sdk/src/airflow/sdk/execution_time/supervisor.py:1]: Okay, reading conde below I understand now that 1 Supervisor :: 1 Task
So Celery --> N Supervisor instances.

File: task_sdk/src/airflow/sdk/execution_time/task_runner.py [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+        buffer = self.buffer
+        self.encoder.encode_into(msg, buffer)
+        buffer += b"\n"
+
+        log.debug("Sending request", json=buffer)
+        self.request_socket.write(buffer)
+
+
+# This global variable will be used by Connection/Variable classes etc to send requests to
+SUPERVISOR_COMMS: CommsDecoder
```
Comment (jscheffl) [inline task_sdk/src/airflow/sdk/execution_time/task_runner.py:107]: Ähm... and how will this work if multiple Supervisor instances are running in parallel?

**Review by jscheffl — COMMENTED** [review]
Wow. This is a large one and probably I need a second pass to understand more. Defintely this is not an easy read.

Is there a way to have a example task execution being "working" with the classes here? (Assume a Celery or similar integration would be a follow-up PR?)

**Comment by raphaelauv** [comment]
do you plan to make `runner` mode customizable at task level ( like executor ) ?

cause I have multiple `operational` pythonoperator that manage heavy database operations like this 

```python
    def clear_dag_runs(dag_id, status_to_clear):
        context = get_current_context()
        session = settings.Session()

        query = session.query(DagRun).filter(
            DagRun.state == status_to_clear, DagRun.dag_id == dag_id)
        rst = query.all()

        dag_bag = DagBag(dag_folder=path.join(SRC_FOLDER, 'dags'), include_examples=False)
        dag: DAG = dag_bag.get_dag(dag_id, session)

        for dag_run in rst:
            dag.clear(
                start_date=dag_run.logical_date,
                end_date=dag_run.logical_date,
                task_ids=None,
                include_subdags=True,
                include_parentdag=True,
                only_failed=False,
            )
        session.close()
```

File: task_sdk/src/airflow/sdk/execution_time/task_runner.py (hunk 11 above)
Comment (ashb) [reply task_sdk/src/airflow/sdk/execution_time/task_runner.py:107]: This runs in the individual task process (forked or subprocess) so there can only ever be one per process.

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py (hunk 10 above)
Comment (ashb) [reply task_sdk/src/airflow/sdk/execution_time/supervisor.py:471]: I considered it briefly, but the total number of these we have is not going to be much more than the comments, so a full on code generator didn't seem worth the cost (overly generic, harder to debug etc.)

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py
Comment (ashb) [reply task_sdk/src/airflow/sdk/execution_time/supervisor.py:1]: Yeah, each celery task will have one supervisor. This Supervisor is roughly the same as LocalTaskJob today.

File: task_sdk/pyproject.toml (hunk 1 above)
Comment (ashb) [reply task_sdk/pyproject.toml:29]: Yes, I've been using it for 2 years without issue across updates.

File: task_sdk/src/airflow/sdk/api/client.py (hunk 2 above)
Comment (ashb) [reply task_sdk/src/airflow/sdk/api/client.py:139]: `airflow task test` equivalent, and an aid to debug and implementation.

File: task_sdk/src/airflow/sdk/api/client.py (hunk 3 above)
Comment (ashb) [reply task_sdk/src/airflow/sdk/api/client.py:66]: Not currently no, good shout, I'll add a ticket to add retry to it.

File: task_sdk/src/airflow/sdk/api/datamodels/_generated.py (hunk 4 above)
Comment (ashb) [reply task_sdk/src/airflow/sdk/api/datamodels/_generated.py:87]: 🤷🏻 - auto generated file based on the OpenAPI spec. We'll look at tweaking these names in the openapi document so these names are better/more stable/predictable soon.

File: task_sdk/src/airflow/sdk/api/datamodels/activities.py (hunk 6 above)
Comment (ashb) [reply task_sdk/src/airflow/sdk/api/datamodels/activities.py:29]: Path to the dag file to parse. I'll rename it to `dag_path` I guess

File: task_sdk/src/airflow/sdk/api/datamodels/ti.py (hunk 7 above)
Comment (ashb) [reply task_sdk/src/airflow/sdk/api/datamodels/ti.py:29]: Yes, likely

File: task_sdk/src/airflow/sdk/execution_time/comms.py (hunk 8 above)
Comment (ashb) [reply task_sdk/src/airflow/sdk/execution_time/comms.py:27]: the Supervisor _could_ be written to be async, but the task subprocess itself can't be async right now as task code is sync. (We'll need to do something async around the triggerer)

The Supervisor is a akin to an HTTP proxy for this traffic, but also has other responsibilities (such as watching the process for exting etc.)

**Comment by ashb** [comment]
> do you plan to make `runner` mode customizable at task level ( like executor ) ?
> 
> cause I have multiple `operational` pythonoperator that manage heavy database operations like this

@raphaelauv No plans today. This is not something in scope for AIP-72 (nor is it possible today)

File: task_sdk/pyproject.toml (hunk 1 above)
Comment (kaxil) [reply task_sdk/pyproject.toml:29]: Yeah it is just one person: https://github.com/jcrist/msgspec/graphs/contributors but shouldn't be tough to replace it with Pydantic in future if needed.

File: task_sdk/pyproject.toml [hunk 12]
```diff
@@ -46,11 +50,24 @@ namespace-packages = ["src/airflow"]
 # Ignore Doc rules et al for anything outside of tests
 "!src/*" = ["D", "TID253", "S101", "TRY002"]
 
-"src/airflow/sdk/__init__.py" = ["TCH004"]
+# Only have pytest rules in tests - https://github.com/astral-sh/ruff/issues/14205
```
Comment (kaxil) [inline task_sdk/pyproject.toml]:
```suggestion
# Only have pytest rules for non-tests code - https://github.com/astral-sh/ruff/issues/14205
```

File: task_sdk/src/airflow/sdk/api/client.py [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+    __slots__ = ("client",)
+
+    def __init__(self, client: Client):
+        self.client = client
+
+    def start(self, id: uuid.UUID, pid: int, when: datetime):
+        """Tell the API server that this TI has started running."""
+        body = TIEnterRunningPayload(pid=pid, hostname=get_hostname(), unixname=getuser(), start_date=when)
+
+        self.client.patch(f"task_instance/{id}/state", content=self.client.encoder.encode(body))
```
Comment (kaxil) [inline task_sdk/src/airflow/sdk/api/client.py]: I changed this recently to `task-instance` but we can handle it in a separate PR

File: task_sdk/src/airflow/sdk/api/client.py [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+        if dry_run:
+            # If dry run is requests, install a no op handler so that simple tasks can "heartbeat" using a
+            # real client, but just don't make any HTTP requests
+            kwargs["transport"] = httpx.MockTransport(noop_handler)
+            kwargs["base_url"] = "dry-run://server"
+        else:
+            kwargs["base_url"] = base_url
+        super().__init__(
+            auth=auth,
+            headers={"airflow-api-version": "2024-07-30"},
```
Comment (kaxil) [inline task_sdk/src/airflow/sdk/api/client.py]: For PR reviewers: To be handled in a separate PR

File: task_sdk/src/airflow/sdk/api/client.py [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+            event_hooks={"response": [raise_on_4xx_5xx], "request": [add_correlation_id]},
```
Comment (kaxil) [inline task_sdk/src/airflow/sdk/api/client.py:159]: I assume this is mainly for tracking each request with id & log

File: task_sdk/src/airflow/sdk/api/client.py [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+                msg = "Remote server returned validation error"
+            else:
+                msg = err.detail.get("message", "") or "Un-parseable error"
+        except Exception:
+            err = msgspec.json.decode(response.content)
+            msg = "Server returned error"
+
+        self = cls(msg, request=response.request, response=response)
+        self.detail = err
+        return self
```
Comment (kaxil) [inline task_sdk/src/airflow/sdk/api/client.py]: Once we start adding more things, it is worth adding tests for things like this func.

Comment (kaxil) [reply task_sdk/src/airflow/sdk/api/client.py]: (in a separate pr)

File: task_sdk/src/airflow/sdk/api/datamodels/_generated.py (hunk 5 above)
Comment (kaxil) [reply task_sdk/src/airflow/sdk/api/datamodels/_generated.py:67]: This doesn't have terminal states like `SUCCESS`, `FAILED` or `SKIPPED`

File: task_sdk/src/airflow/sdk/api/datamodels/_generated.py (hunk 4 above)
Comment (kaxil) [reply task_sdk/src/airflow/sdk/api/datamodels/_generated.py:87]: Yeah, I will fix it in a separate PR

File: task_sdk/src/airflow/sdk/api/datamodels/ti.py [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+    run_id: str
+    try_number: int
+    map_index: Optional[int] = None  # noqa: UP007 msgspec doesn't support pipe-op
```
Comment (kaxil) [inline task_sdk/src/airflow/sdk/api/datamodels/ti.py]: Should we default it to `-1`, or not make any such assumptions in the API?

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py [hunk 18]
```diff
@@ -0,0 +1,552 @@
...
+        try:
+            last_chance_stderr.write("--- Last chance exception handler ---\n")
+            traceback.print_exception(exc, value=v, tb=tb, file=last_chance_stderr)
+            exit(99)
+        except Exception as e:
+            with suppress(Exception):
+                print(
+                    f"--- Last chance exception handler failed --- {repr(str(e))}\n", file=last_chance_stderr
+                )
+            exit(98)
```
Comment (kaxil) [inline task_sdk/src/airflow/sdk/execution_time/supervisor.py]: Does exit code `98` and `99` represent something? or we use it to handle the logic somewhere?

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py [hunk 19]
```diff
@@ -0,0 +1,552 @@
...
+    main()
+
+
+def _fork_main(
+    child_stdin: socket,
+    child_stdout: socket,
+    child_stderr: socket,
+    log_fd: int,
+    target: Callable[[], None],
+) -> NoReturn:
```
Comment (kaxil) [inline task_sdk/src/airflow/sdk/execution_time/supervisor.py]: Worth adding docstring with the tldr of what things are done in this func

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py [hunk 20]
```diff
@@ -0,0 +1,552 @@
...
+
+        try:
+            while self._exit_code is None or len(self.selector.get_map()):
+                last_heartbeat_ago = time.monotonic() - self._last_heartbeat
+                # Monitor the task to see if it's done. Wait in a syscall (`select`) for as long as possible
+                # so we notice the subprocess finishing as quick as we can.
+                max_wait_time = max(
+                    0,  # Make sure this value is never negative,
+                    min(
+                        # Ensure we heartbeat _at most_ 75% through the time the zombie threshold time
```
Comment (kaxil) [inline task_sdk/src/airflow/sdk/execution_time/supervisor.py]:
```suggestion
                        # Ensure we heartbeat _at most_ 75% through the the zombie threshold time
```

File: task_sdk/src/airflow/sdk/log.py [hunk 21]
```diff
@@ -0,0 +1,377 @@
...
+        import contextlib
+
+        import click
+        import httpcore
+        import httpx
+        import typer
+
+        rich_exc_formatter = structlog.dev.RichTracebackFormatter(
+            extra_lines=0,
+            max_frames=30,
```
Comment (kaxil) [inline task_sdk/src/airflow/sdk/log.py:164]: Any reason for 0 and 30 here than the defaults?

File: task_sdk/src/airflow/sdk/log.py [hunk 22]
```diff
@@ -0,0 +1,377 @@
...
+
+        encoder = msgspec.json.Encoder()
+
+        def json_dumps(msg, default):
+            return encoder.encode(msg)
+
+        def json_processor(logger: Any, method_name: Any, event_dict: EventDict) -> str:
+            # import web_pdb
+
+            # web_pdb.set_trace()
```
Comment (kaxil) [inline task_sdk/src/airflow/sdk/log.py]:
```suggestion
```

File: task_sdk/tests/execution_time/test_supervisor.py [hunk 23]
```diff
@@ -0,0 +1,150 @@
...
+import os
+import signal
+import sys
+from unittest.mock import MagicMock
+
+import pytest
+import structlog
+import structlog.testing
+
+import airflow.sdk.api.client
```
Comment (kaxil) [inline task_sdk/tests/execution_time/test_supervisor.py]:
```suggestion
from airflow.sdk.api import client
```

Comment (kaxil) [reply task_sdk/tests/execution_time/test_supervisor.py]: or was it on purpose?

**Review by kaxil — APPROVED** [review]
lgtm. Worth splitting certain methods/func into more granular funcs and adding more docstrings wherever you can.

**Comment by ashb** [comment]
> lgtm. Worth splitting certain methods/func into more granular funcs and adding more docstrings wherever you can.

For sure, I will do that before merging and ping you tomrorow for another look

File: task_sdk/tests/execution_time/test_supervisor.py (hunk 23 above)
Comment (ashb) [reply task_sdk/tests/execution_time/test_supervisor.py]: Probably on purpose of "`Client` on it's own is a big ambigious what it is"

File: task_sdk/src/airflow/sdk/log.py (hunk 22 above)
Comment (ashb) [reply task_sdk/src/airflow/sdk/log.py]: Whoops 😁

File: task_sdk/src/airflow/sdk/log.py (hunk 21 above)
Comment (ashb) [reply task_sdk/src/airflow/sdk/log.py:164]: Likely not! Picked largely at random. `extra_lines=0` was to make the stack traces smaller (as otherwise they get huge fast). max frames was an arbitrary choice

File: task_sdk/src/airflow/sdk/api/client.py (hunk 15 above)
Comment (ashb) [reply task_sdk/src/airflow/sdk/api/client.py:159]: Correct yes.

File: task_sdk/pyproject.toml (hunk 1 above)
Comment (jscheffl) [reply task_sdk/pyproject.toml:29]: Just wanted to ask if we have experience with this... not that we need to "clean fresh dirt on the beach" :-D

File: task_sdk/src/airflow/sdk/api/client.py (hunk 2 above)
Comment (jscheffl) [reply task_sdk/src/airflow/sdk/api/client.py:139]: Can you add a comment for this... nor other noops like me coming around or my future me? :-D

File: task_sdk/src/airflow/sdk/api/datamodels/ti.py (hunk 7 above)
Comment (jscheffl) [reply task_sdk/src/airflow/sdk/api/datamodels/ti.py:29]: In another / Other/future PR? Can you leave a TODO note in the code then? (maybe with a link to a ticket)

File: task_sdk/src/airflow/sdk/execution_time/comms.py (hunk 8 above)
Comment (jscheffl) [reply task_sdk/src/airflow/sdk/execution_time/comms.py:27]: Okay so then this is a direct proxy-ing - if upload line bandwidth is limited, this would then limit the performance of task execution as all log messages are directly "write though" as stream? Or is there some buffering (would increase performance but potentially lose messages on crash?)

File: task_sdk/src/airflow/sdk/api/client.py (hunk 14 above)
Comment (jscheffl) [reply task_sdk/src/airflow/sdk/api/client.py]: Why not leaving a TODO with a comment in code? :-D

File: task_sdk/src/airflow/sdk/api/client.py (hunk 16 above)
Comment (jscheffl) [reply task_sdk/src/airflow/sdk/api/client.py]: Can you leave a TODO then here as reminder than this is considered and not done yet?

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py (hunk 18 above)
Comment (jscheffl) [reply task_sdk/src/airflow/sdk/execution_time/supervisor.py]: That was the signal in BashOperator (back long time ago) to mark a task as skipped.

File: task_sdk/tests/execution_time/test_supervisor.py (hunk 23 above)
Comment (jscheffl) [reply task_sdk/tests/execution_time/test_supervisor.py]: How about?
```suggestion
from airflow.sdk.api import client as sdk_client
```

File: task_sdk/src/airflow/sdk/api/client.py (hunk 14 above)
Comment (kaxil) [reply task_sdk/src/airflow/sdk/api/client.py]: Absolutely :)

File: task_sdk/src/airflow/sdk/execution_time/comms.py (hunk 8 above)
Comment (ashb) [reply task_sdk/src/airflow/sdk/execution_time/comms.py:27]: Log messages aren't currently handled in this PR -- or at least they aren't being sent upstream to the API server, log messages are just being collected from the subprocess (as json) and then set to stdout of the Supervisor.

"shortly"/in the next PR all those logs will be written to local disk on the worker in files as they do today, and _then_ we can see about improving things for logs to S3 and of running tasks etc.

File: task_sdk/tests/execution_time/test_supervisor.py (hunk 23 above)
Comment (ashb) [reply task_sdk/tests/execution_time/test_supervisor.py]: Yeah that works

File: task_sdk/src/airflow/sdk/api/datamodels/ti.py (hunk 17 above)
Comment (ashb) [reply task_sdk/src/airflow/sdk/api/datamodels/ti.py]: The main reason we have -1 in the server side code is that since map_index is part of the primary key it can't be nullable. Null is conceptually a better indicator of "I'm not mapped"

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py (hunk 18 above)
Comment (ashb) [reply task_sdk/src/airflow/sdk/execution_time/supervisor.py]: Oh was it? Oops, I should change those then. These are just meant to be a marker of "hey, something _really_ odd happened in this process"

Comment (ashb) [reply task_sdk/src/airflow/sdk/execution_time/supervisor.py]: Changed to 126 and 125, and added a comment.

File: task_sdk/src/airflow/sdk/api/client.py (hunk 14 above)
Comment (ashb) [reply task_sdk/src/airflow/sdk/api/client.py]: I've removed this for now (and replaced it with a user agent string of `apache-airflow-task-sdk/1.0.0dev0 (Python/3.12.7)` (for example)

File: task_sdk/tests/execution_time/test_supervisor.py (hunk 23 above)
Comment (ashb) [reply task_sdk/tests/execution_time/test_supervisor.py]: Done

File: task_sdk/src/airflow/sdk/log.py (hunk 21 above)
Comment (ashb) [reply task_sdk/src/airflow/sdk/log.py:164]: Left a comment saying as much

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py (hunk 19 above)
Comment (ashb) [reply task_sdk/src/airflow/sdk/execution_time/supervisor.py]: Doc string, and refactor to small name functions to make flow a bit clearer hopefully https://github.com/apache/airflow/pull/43893/commits/c54c51e8257f87410267aa5da636a326a9028089#diff-c2651fdee1a25e091e2a9d4f937f8032ca3d289d0de76f38ed88aee5df0f880dR159-R183

File: task_sdk/src/airflow/sdk/api/client.py [hunk 24]
```diff
@@ -0,0 +1,218 @@
...
+        self.client.patch(f"task_instance/{id}/state", content=self.client.encoder.encode(body))
+
+    def finish(self, id: uuid.UUID, state: TaskInstanceState, when: datetime):
+        """Tell the API server that this TI has reached a terminal state."""
+        body = TITerminalStatePayload(end_date=when, state=TerminalState(state))
+
+        self.client.patch(f"task_instance/{id}/state", content=self.client.encoder.encode(body))
+
+    def heartbeat(self, id: uuid.UUID):
+        self.client.put(f"task_instance/{id}/heartbeat")
```
Comment (kaxil) [inline task_sdk/src/airflow/sdk/api/client.py]:
```suggestion
        self.client.patch(f"task-instance/{id}/state", content=self.client.encoder.encode(body))

    def heartbeat(self, id: uuid.UUID):
        self.client.put(f"task-instance/{id}/heartbeat")
```

File: task_sdk/src/airflow/sdk/api/client.py [hunk 25]
```diff
@@ -0,0 +1,218 @@
...
+    encoder: msgspec.json.Encoder
+
+    def __init__(self, *, base_url: str | None, dry_run: bool = False, token: str, **kwargs: Any):
+        if (not base_url) ^ dry_run:
+            raise ValueError(f"Can only specify one of {base_url=} or {dry_run=}")
+        auth = BearerAuth(token)
+
+        self.encoder = msgspec.json.Encoder()
+        if dry_run:
+            # If dry run is requests, install a no op handler so that simple tasks can "heartbeat" using a
```
Comment (kaxil) [inline task_sdk/src/airflow/sdk/api/client.py]: nit: "If dry run is requested" or "If the request is dry-run"

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py [hunk 26]
```diff
@@ -0,0 +1,589 @@
...
+            # Ensure the buffer doesn't grow and stay large if a large payload is used. This won't grow it
+            # larger than it is, but it will shrink it
+            if len(buffer) > 1024:
+                buffer = buffer[:1024]
+
+
+# Sockets, even the `.makefile()` function don't correctly do line buffering on reading. If a chunk is read
+# and it doesn't contain a new line character, `.readline()` will just return the chunk as is.
+#
+# This returns a cb suitable for attaching to a `selector` that reads in to a buffer, and yields lines to a
```
Comment (kaxil) [inline task_sdk/src/airflow/sdk/execution_time/supervisor.py]: `cb` shortform of `callback`?

Comment (kaxil) [reply task_sdk/src/airflow/sdk/execution_time/supervisor.py]: Yes, based on the code

File: task_sdk/src/airflow/sdk/log.py [hunk 27]
```diff
@@ -0,0 +1,371 @@
...
+        import httpx
+        import typer
+
+        dict_exc_formatter = structlog.tracebacks.ExceptionDictTransformer(
+            use_rich=False, show_locals=False, suppress=(click, typer)
+        )
+
+        dict_tracebacks = structlog.processors.ExceptionRenderer(
+            structlog.tracebacks.ExceptionDictTransformer(
+                use_rich=False, show_locals=False, suppress=(click, typer)
```
Comment (kaxil) [inline task_sdk/src/airflow/sdk/log.py]: Do we need to add `asyncio, httpcore, httpx, contextlib` too or remove the imports in the `else` clause?

**Review by kaxil — APPROVED** [review]
small nits, more tests and we are good to go

File: task_sdk/src/airflow/sdk/api/client.py (hunk 16 above)
Comment (ashb) [reply task_sdk/src/airflow/sdk/api/client.py]: Added tests for this now

File: task_sdk/pyproject.toml (hunk 12 above)
Comment (ashb) [reply task_sdk/pyproject.toml]: Changed to say `# Ignore the pytest rules outside the tests folder - https://github.com/astral-sh/ruff/issues/14205
`

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py (hunk 26 above)
Comment (ashb) [reply task_sdk/src/airflow/sdk/execution_time/supervisor.py]: Yes, Changed

File: task_sdk/src/airflow/sdk/log.py (hunk 27 above)
Comment (ashb) [reply task_sdk/src/airflow/sdk/log.py]: Yeah updated.

File: task_sdk/src/airflow/sdk/execution_time/comms.py (hunk 9 above)
Comment (ashb) [reply task_sdk/src/airflow/sdk/execution_time/comms.py:38]: Created issue for us https://github.com/orgs/apache/projects/405/views/1?pane=issue&itemId=87109558

File: task_sdk/src/airflow/sdk/api/client.py (hunk 3 above)
Comment (ashb) [reply task_sdk/src/airflow/sdk/api/client.py:66]: https://github.com/orgs/apache/projects/405/views/1?pane=issue&itemId=87109558

File: task_sdk/tests/execution_time/test_task_runner.py [hunk 28]
```diff
@@ -0,0 +1,57 @@
...
+        assert msg.ti.id == uuid.UUID("4d828a62-a417-4936-a7a6-2b3fabacecab")
+        assert msg.ti.task_id == "a"
+        assert msg.ti.dag_id == "c"
+        assert msg.file == "/dev/null"
+
+        # Since this was a StartupDetails message, the decoder should open the other socket
+        assert decoder.request_socket.writable()
+        assert decoder.request_socket.fileno() == w2.fileno()
+
+        ...
```
Comment (kaxil) [inline task_sdk/tests/execution_time/test_task_runner.py]: ?

Comment (ashb) [reply task_sdk/tests/execution_time/test_task_runner.py]: Whops, left over.

File: task_sdk/src/airflow/sdk/api/datamodels/ti.py (hunk 17 above)
Comment (jscheffl) [reply task_sdk/src/airflow/sdk/api/datamodels/ti.py]: Like that idea. Was always bothered with the -1 ... but now I see it was not a badly introduced legacy but primary key is a reason I did not think of so far :-D

**Review by jscheffl — APPROVED** [review]
